rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+    The nested structure, `Tensor` shapes, and `source_replica_id` must be
+    identical in all replicas.
+
+    IMPORTANT: The ordering of communications must be identical in all replicas.
+
+    `value` will always be evaluated on all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}, `source_replica_id`: 1
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}, `source_replica_id`: 1
+
+      Replica 0:: result: {'a': 3, 'b': [ 2, 98]}
+      Replica 1:: result: {'a': 3, 'b': [ 2, 98]}
+
+    Args:
+      value: The nested structure of `Tensor`s to broadcast.
+        The structure must be compatible with `tf.nest`.
+      source_replica_id: The ID of the replica from which to broadcast.
+
+    Returns:
+      A `Tensor` nest with the `value` from the source replica.
+    """
+```
+
+#### InputReplicationMode
+
+```python
+class InputReplicationMode(enum.Enum):
+  SINGLE = 0
+  PER_WORKER = 1
+  PER_REPLICA = 2
+```
+
+##### SINGLE
+
+The input function will be called once, typically on the first worker. The
+entire input pipeline ops then live on that worker. All replicas (on that and
+other workers) will dequeue from a single `Dataset` created on that worker.
+
+This mode is not supported by between-graph Distribution Strategy
+implementations.
+
+##### PER_WORKER
+
+The input function will be called on each worker independently, creating as many
+input pipelines as number of workers. Replicas will dequeue from the local
+`Dataset` on their worker. Replicator doesn't manage any state sharing between
+such separate input pipelines.
+
+##### PER_REPLICA
+
+The input function will be called for every replica, on the corresponding
+worker, thus creating as many input pipelines as number of replicas. Each
+replica will dequeue from its own `Dataset`.
+
+#### InputContext
+
+An instance of `InputContext` is passed to the input function.
+
+```python
+class InputContext(object):
+  @property
+  def num_replicas(self) -> int:
+    """Returns the total number of compute replicas."""
+
+  @property
+  def num_replicas_in_sync(self) -> int:
+    """Returns the number of compute replicas training in sync."""
+
+  @property
+  def input_pipeline_id(self) -> int:
+    """Returns the input pipeline ID.
+
+    If `replication_mode` == `SINGLE`, this is always `0`.
+    If `replication_mode` == `PER_WORKER`, this is the worker ID.
+    If `replication_mode` == `PER_REPLICA`, this is the compute replica ID.
+    """
+
+  @property
+  def num_input_pipelines(self) -> int:
+    """Returns the number of input pipelines.
+
+    If `replication_mode` == `SINGLE`, this is always `1`.
+    If `replication_mode` == `PER_WORKER`, this is the number of workers.
+    If `replication_mode` == `PER_REPLICA`, this is `num_replicas`.
+    """
+```
+
+#### InputIterator
+
+An instance of `InputIterator` is returned by `prepare_input` and can be passed
+to `run`.
+
```

Comment:
Why is it necessary to have this additional wrapping around the Dataset? Just to handle the non-Dataset case? Or does it serve another purpose?

Reply:
Currently it does server another purpose. It allows us to call the function on different workers and create multiple datasets (for the PER_WORKER mode). So each worker will have their own copy of the dataset. If the user gave us an already created dataset, we have to send the data from one worker to all the other workers, and this is not efficient currently. So we support the PER_WORKER mode and ask this to be a function.
That being said, there is a proposal in the works for a distributed dataset which will alleviate this sort of requirement and allow us to efficiently distribute from a single dataset.